bowtie/_cli.py

```diff
@@ -1684,6 +1708,623 @@ async def info(
+                ],
+            )
+
+            progress.update(
+                task,
+                description=(
+                    f"Successfully downloaded and parsed all latest "
+                    f"reports of {id} for {pretty_names_str}!"
+                ),
+                completed=total_files,
+                total=total_files,
+            )
+            return responses
+
+
+def _trend_table_for(
+    id: ConnectableId,
+    versions: Set[str],
+    dialects_trend: dict[Dialect, _report.Report],
+) -> Table:
+    main_table = Table(show_lines=True)
+    main_table.add_column(
+        "Dialect",
+        justify="center",
+        vertical="middle",
+    )
+    main_table.add_column(
+        f"Trend Data of {id} versions",
+        justify="center",
+    )
+
+    def add_sub_table_row(
+        sub_table: Table,
+        version: str,
+        unsuccessful: Unsuccessful,
+    ):
+        sub_table.add_row(
+            version,
+            str(len(unsuccessful.skipped)),
+            str(len(unsuccessful.errored)),
+            str(len(unsuccessful.failed)),
+        )
+
+    for dialect, report in dialects_trend.items():
+        test = "tests" if report.total_tests != 1 else "test"
+        sub_table = Table(
+            "Version",
+            "Skips",
+            "Errors",
+            "Failures",
+            caption=f"{report.total_tests} {test} ran",
+            show_edge=False,
+        )
+
+        if versions:
+            for version, unsuccessful in report.latest_to_oldest():
+                add_sub_table_row(sub_table, version, unsuccessful)
+        else:
+            implementation = report.implementations[id]
+            version = implementation.version or "latest"
+            unsuccessful = report.unsuccessful(id)
+            add_sub_table_row(sub_table, version, unsuccessful)
+
+        main_table.add_row(dialect.pretty_name, sub_table)
+
+    return main_table
+
+
+def _trend_table_in_markdown_for(
+    id: ConnectableId,
+    versions: Set[str],
+    dialects_trend: dict[Dialect, _report.Report],
+) -> str:
+    rows_data: list[list[str]] = []
+
+    inner_table_columns = [
+        "Version",
+        "Skips",
+        "Errors",
+        "Failures",
+    ]
+
+    def create_inner_table_row(
+        version: str,
+        unsuccessful: Unsuccessful,
+    ):
+        return [
+            version,
+            str(len(unsuccessful.skipped)),
+            str(len(unsuccessful.errored)),
+            str(len(unsuccessful.failed)),
+        ]
+
+    for dialect, report in dialects_trend.items():
+        test = "tests" if report.total_tests != 1 else "test"
+        inner_table_rows: list[list[str]] = []
+        if versions:
+            for version, unsuccessful in report.latest_to_oldest():
+                inner_table_rows.append(
+                    create_inner_table_row(version, unsuccessful),
+                )
+        else:
+            implementation = report.implementations[id]
```

Comment:
This seems like again it should be handled somewhere earlier -- if there's no versions, make wherever you get `versions` from just contain the one version.

Reply:
> This seems like again it should be handled somewhere earlier -- if there's no versions, make wherever you get versions from just contain the one version.

I think the solution for this would be better served once we manage to get PR #1487 merged. Since after that `get_versions` would simply be available on `Implementation` itself and then that will be single handedly responsible for returning the correct `versions` data back and then I won't have to deal with just just `ConnectableId`s for `bowtie trend` (cause currently that's how I have set this up (`bowtie trend` is not really starting an implementation at all right now)).